Annotation deletion popup (bug 1899731)

ryzokuken opened the pull request:
This PR adds a simple popup/snackbar setup and wires it up for annotation removal, so when a user deletes any number of annotations, they get a message about it alongside an undo button that allows them to quickly undo the deletion.

ryzokuken committed:
Break messageBar class out

Move .messageBar out of .dialog into its own standalone class in order
to reuse as much of it for the upcoming feature for an undo message for
annotations.

nicolo-ribaudo reviewed (commented):
This will need tests in the various editors, in https://github.com/mozilla/pdf.js/tree/master/test/integration, or maybe its own new test file that tests that the toast is properly shown in the various cases.

nicolo-ribaudo commented on web/app.js:
```diff
@@ -450,6 +451,10 @@ const PDFViewerApplication = {
           )
         : null;
     }
+    const toastManager = new ToastManager(
+      appConfig.annotationRemovedToast,
+      10000
+    );
```
```js
    const toastManager = new ToastManager({
      element: appConfig.annotationRemovedToast,
      timeout: 10000
    });
```

some API like this would make it easier to understand what "10000" means when reading the code.

calixteman replied:
Has this value been specified ?

Snuffleupagus replied:
This will break in e.g. the GeckoView viewer, note how other definitions handle this via `appConfig`-checks in this code.

ryzokuken replied:
Thanks for the suggestions and to be entirely fair I had no clue what was a good thing to do in this specific case so I decided to go with something mostly arbitrary and ask you all for guidance whether the project's guidelines have something to say on this. Thanks for the pointer @Snuffleupagus, I'll look around the codebase for what to do here.

ryzokuken replied:
@Snuffleupagus after a quick look, I could find most current calls to `setInterval` in the codebase atm either with hardcoded values or a literal constant defined in the respective file, could you point me to the part of the codebase you're specifically referring to above? Thanks

Snuffleupagus replied:
*(I don't know what setInterval has to do with this, since that's not related to the issue I mentioned above.)*

Following the `viewer.js` changes, this is safe if implemented as:
```suggestion
    const editorUndoBar = appConfig.editorUndoBar
      ? new EditorUndoBar(appConfig.editorUndoBar, /* duration = */ 10000)
      : null;
```

Snuffleupagus reviewed (changes requested):
Is this perhaps related to https://bugzilla.mozilla.org/show_bug.cgi?id=1899731, since there seems to be no mention of that bug here?

Please note that we'll need a "full" design specification *before* implementing this, and given its current state the code will also require a fair amount of work here (as far as I can tell).

nicolo-ribaudo commented on web/message_bar.css:
```diff
@@ -0,0 +1,153 @@
+		font-weight: bold;
+		color: #15141a;
+		border: none;
+    padding: 8px 16px;
```
Formatting

calixteman commented on web/viewer.html:
```diff
@@ -686,6 +686,18 @@
+              <span class="description" data-l10n-id="pdfjs-editor-messagebar-annotation-removed-text" data-l10n-args='{"type": "annotation"}'>[Annotation] removed</span>
+            </div>
+            <button id="annotationRemovedUndoButton" class="undoButton" type="button" tabindex="0" title="Undo"><span data-l10n-id="pdfjs-editor-messagebar-annotation-undo-button">Undo</span></button>
+            <button id="annotationRemovedCloseButton" class="closeButton" type="button" tabindex="1" title="Close"><span data-l10n-id="pdfjs-editor-messagebar-annotation-close-button">Close</span></button>
```
Don't use `tabindex=X` with `X > 0`.

ryzokuken replied:
Thanks for the pointer, I was totally unsure of how `tabIndex` works to begin with :sweat_smile: 

I've switched it with `0`, hope that's what was needed. I tested and the button is focusable.

Snuffleupagus commented on web/message_bar.css:
This is missing the licence header at the top, see the other CSS files.

Snuffleupagus replied:
This wasn't fixed in the correct patch, since the change belongs in the `Break messageBar class out`.

ryzokuken replied:
I intend to squash it all together since it's a related change unless you prefer smaller commits instead

calixteman commented on src/display/editor/tools.js:
```diff
@@ -2002,6 +2006,12 @@ class AnnotationEditorUIManager {
+        undo,
+        editors.length > 1
+          ? `${editors.length} annotations`
+          : editors[0].constructor._type
```
You can use `editors[0].editorType`.

Snuffleupagus commented on web/message_bar.css:
```diff
@@ -0,0 +1,119 @@
+  width: 400px;
+  left: 50%;
+  transform: translateX(-50%);
+}
```
This shouldn't be included in this patch, but the in the next one.

Snuffleupagus replied:
This hasn't been fixed in the correct patch, since the change belongs in `Break messageBar class out`.

Snuffleupagus commented on web/pdf_viewer.css:
```diff
@@ -16,6 +16,7 @@
 @import url(text_layer_builder.css);
 @import url(annotation_layer_builder.css);
 @import url(xfa_layer_builder.css);
+@import url(message_bar.css);
```
Would this not make more sense to place *before* the `@import url(dialog.css);` line, since that file extends the `.messageBar` class?

Snuffleupagus replied:
This wasn't fixed in the correct patch, since the change belongs in `Break messageBar class out`.

calixteman commented on web/message_bar.css:
```diff
@@ -0,0 +1,153 @@
+	bottom: 100px;
+	left: 50%;
+	transform: translateX(-50%);
+}
```
Did you test that everything is rendering fine when switching to HCM mode on Windows ?

ryzokuken replied:
Sorry I didn't respond to this comment earlier, it has indeed been tested on Win11 and it works as expected.

Snuffleupagus commented on l10n/en-US/viewer.ftl:
```diff
@@ -493,3 +493,9 @@ pdfjs-editor-alt-text-settings-editor-title = Alt text editor
+#   $type (String) - the type of annotation that was just removed with an optional quantity.
+pdfjs-editor-messagebar-annotation-removed-text = { $type } removed
+pdfjs-editor-messagebar-annotation-undo-button = Undo
+pdfjs-editor-messagebar-annotation-close-button = Close
```
Missing new-line here.

Snuffleupagus commented on src/display/editor/tools.js:
```diff
@@ -2002,6 +2006,12 @@ class AnnotationEditorUIManager {
+      this.#toastManager.show(
+        undo,
+        editors.length > 1
+          ? `${editors.length} annotations`
```
This won't be possible to localize properly.

Snuffleupagus commented on web/message_bar.css:
```diff
@@ -0,0 +1,153 @@
+
+#annotationRemovedToast {
+  --message-bar-icon: url(images/secondaryToolbarButton-documentProperties.svg);
+  --message-bar-icon-color: #0060DF;
+	--message-bar-bg-color: #deeafc;
+	--message-bar-fg-color: #15141a;
+	--text-primary-color: #15141a;
```
Should this not be using different colours depending on light/dark/HCM mode?

Snuffleupagus commented on web/pdf_viewer.js:
```diff
@@ -203,6 +203,8 @@ class PDFViewer {
 
   #altTextManager = null;
 
+  #toastManager = null;
```
Here, and in lots of other spots in the patch, please place class fields in *alphabetical* order.

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,52 @@
+    }, this.#duration);
+    this.#controller = new AbortController();
+    console.log(this.#toastElement, this.#toastElement.getElementById);
+    this.#toastElement
+      .querySelector("#annotationRemovedUndoButton")
```
Here, and elsewhere, please note that you're not "allowed" to lookup elements like this and compare with how e.g. the toolbars are handled.

Snuffleupagus commented on web/viewer.js:
```diff
@@ -223,6 +223,7 @@ function getViewerConfiguration() {
       editorHighlightShowAll: document.getElementById("editorHighlightShowAll"),
     },
     printContainer: document.getElementById("printContainer"),
+    annotationRemovedToast: document.getElementById("annotationRemovedToast"),
```
This should be handled similar to the toolbars, see above.

Snuffleupagus replied:
You want something like the following:
```suggestion
    editorUndoBar: {
      container: document.getEleementById(...);
      undoButton: document.getEleementById(...);
      closeButton: document.getEleementById(...);
    },
```

Snuffleupagus reviewed (changes requested):
Leaving just a few high-level comments regarding things/patterns that should be improved.

nicolo-ribaudo commented on web/toast_manager.js:
```diff
@@ -7,17 +22,24 @@ class ToastManager {
-    this.#timeoutID = setTimeout(() => {
-      this.#hide();
-    }, this.#duration);
+    this.#boundHide = this.#hide.bind(this);
```
This can be moved to the initialization of the field, or to the constructor.

ryzokuken commented:
First off, thanks everyone for all the comments it's nice to have a stronger sense of direction as to where I need to take this patch. I addressed some of the comments y'all made and marked the appropriate suggestions as "resolved" although please feel free to re-raise anything.

To respond to the questions:

> Is this perhaps related to [bugzilla.mozilla.org/show_bug.cgi?id=1899731](https://bugzilla.mozilla.org/show_bug.cgi?id=1899731), since there seems to be no mention of that bug here?

> Please note that we'll need a "full" design specification before implementing this, and given its current state the code will also require a fair amount of work here (as far as I can tell).

Yes it is! Thanks @marco-c for the clarification and sharing the spec. Regarding the code, I hope the bigger pain points have either already been addressed or I'd be hacking on them shortly.

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+ * limitations under the License.
+ */
+
+class ToastManager {
```
I'm not convinced this is a great name, which applies everywhere through the PR, since this only handles AnnotationEditor undo and nothing else really.

Maybe `EditorUndoBar` instead, since that seems more appropriate?

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+ */
+
+class ToastManager {
+  #toastElement;
+
+  #duration;
+
+  #timeoutID = null;
+
+  #controller = null;
+
```
As already mentioned previously, please sort class fields alphabetically.

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+
+  #boundHide = this.#hide.bind(this);
+
+  constructor(elem, duration) {
```
```suggestion
  constructor({ container, undoButton, closeButton }, duration) {
```

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+  #boundHide = this.#hide.bind(this);
+
+  constructor(elem, duration) {
+    this.#toastElement = elem;
```
```suggestion
    this.#container = container;
    this.#undoButton = undoButton;
    this.#closeButton = closeButton;
```

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+
+  #duration;
+
+  #timeoutID = null;
```
```suggestion
  #timeoutId = null;
```

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+      });
+  }
+
+  #finalizeTimeout() {
+    clearTimeout(this.#timeoutID);
+    this.#timeoutID = null;
+  }
+
+  #finalizeController() {
+    this.#controller.abort();
+    this.#controller = null;
+  }
+
+  #hide() {
+    this.#toastElement.setAttribute("hidden", "");
+    this.#finalizeTimeout();
+    this.#finalizeController();
+  }
```
Why do we need separate methods, since that seems unnecessary?
```suggestion
  #hide() {
    this.#container.hidden = true;
    
    if (this.#timeoutId) {
      clearTimeout(this.#timeoutId);
      this.#timeoutId = null;
    }
    this.#controller?.abort();
    this.#controller = null;
  }
```

ryzokuken replied:
They're necessary because I'm also conditionally calling them at the start of `show` to cleanup any previous timeouts or controllers if they exist.

ryzokuken replied:
(requested in https://github.com/mozilla/pdf.js/pull/18900#discussion_r1799764474)

Snuffleupagus replied:
> They're necessary because I'm also conditionally calling them at the start of `show` to cleanup any previous timeouts or controllers if they exist.

Please just implement this, see also https://github.com/mozilla/pdf.js/pull/18900#discussion_r1800676358 below.

ryzokuken replied:
Fixed, thanks I'd missed the other comment

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  show(action, type) {
+    if (this.#timeoutID) {
+      this.#finalizeTimeout();
+    }
+    if (this.#controller) {
+      this.#finalizeController();
+    }
```
```suggestion
    this.#hide();
    
```

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+    this.#toastElement.removeAttribute("hidden");
+    this.#timeoutID = setTimeout(this.#boundHide, this.#duration);
+    this.#controller = new AbortController();
+    this.#toastElement
+      .querySelector("#annotationRemovedUndoButton")
+      .addEventListener(
```
```suggestion
    this.#undoButton.addEventListener(
```

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+    if (this.#controller) {
+      this.#finalizeController();
+    }
+    this.#toastElement.setAttribute("data-l10n-args", JSON.stringify({ type }));
+    this.#toastElement.removeAttribute("hidden");
+    this.#timeoutID = setTimeout(this.#boundHide, this.#duration);
+    this.#controller = new AbortController();
```
```suggestion
    this.#container.setAttribute("data-l10n-args", JSON.stringify({ type }));
    this.#container.hidden = false;
    
    this.#timeoutId = setTimeout(this.#boundHide, this.#duration);
    this.#controller = new AbortController();
    const opts = { signal: this.#controller.signal };
    
```

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+          action();
+          this.#hide();
+        },
+        { signal: this.#controller.signal }
```
```suggestion
        opts
```

Snuffleupagus commented on web/toast_manager.js:
```diff
@@ -0,0 +1,77 @@
+        },
+        { signal: this.#controller.signal }
+      );
+    this.#toastElement
+      .querySelector("#annotationRemovedCloseButton")
+      .addEventListener("click", this.#boundHide, {
+        signal: this.#controller.signal,
+      });
```
```suggestion
    this.#closeButton.addEventListener("click", this.#boundHide, opts);
```

Snuffleupagus reviewed (changes requested):
Based on looking at the code, I believe that there's various cases where the notification won't be closed *immediately* as I believe is intended when:
 - Restoring the editor with <kbd>Ctrl</kbd>+<kbd>Z</kbd> (i.e. using the keyboard).
 - Toggling editing-mode off.
 - The entire editorLayer is being destroyed (i.e. when closing the current PDF document).

ryzokuken commented:
@Snuffleupagus 

> (I don't know what setInterval has to do with this, since that's not related to the issue I mentioned above.)

Apologies, I misunderstood your comment at first. Applied the fix that you'd ultimately suggested.

I've marked a number of your comments as resolved, please take a look and let me know if I missed something from my latest commit while I work on addressing the remaining issues. Thanks again for your patience.

Snuffleupagus commented on web/viewer.html:
```diff
@@ -686,6 +686,18 @@
 <!--#endif-->
       </div>  <!-- dialogContainer -->
 
+      <div id="toastContainer">
```
Here, and elsewhere in the patch, please update this name as well.

ryzokuken replied:
I was actually curious if we should remove the div entirely since it seems that we're not going to focus much on generalizing for future popups like this, what is your preferred name for this div? `editorUndoBarContainer` seems to be a mouthful but I'm more than happy to accept any suggestions.

Snuffleupagus replied:
That seems like a good idea to me. (We can always extend the functionality in the future, if actually needed, but let's keep things as simple as possible for now.)

Snuffleupagus commented on web/message_bar.css:
```diff
@@ -0,0 +1,174 @@
+}
+
+#editorUndoBar {
+	--message-bar-icon: url(images/secondaryToolbarButton-documentProperties.svg);
```
That SVG file seems like a strange choice here, too much copy-paste?

ryzokuken replied:
It was just the closest to the spec that I could find in the existing codebase :sweat_smile: 

I should replace it with a new appropriate SVG but I'd been putting it off until now. I considered downloading the SVG directly from Figma and using that but it has some issues too. I'll ask the UX folks if they have any suggestions here.

ryzokuken replied:
Verified with the folks from the design team that it's a close enough icon to use and the reusability makes it an even nicer choice.

Snuffleupagus commented on web/message_bar.css:
```diff
@@ -0,0 +1,174 @@
+ */
+
+.messageBar {
+	--closing-button-icon: url(images/messageBar_closingButton.svg);
```
The formatting looks "off" all throughout this file, since it seems that you're using tabs (which we don't use anywhere in the code-base).

Running `gulp lint --fix` locally should help fix that, but I'd recommend using an editor that supports `.editorconfig`-files to help avoid that kind of issue.

ryzokuken replied:
I've been using an extension that picks up `.editorconfig` but it was misbehaving for some reason and another formatter was overriding the settings, I'll look into it. In the meantime, thanks for the heads up I'll run `gulp lint --fix` before future commits.

ryzokuken commented on web/viewer.html:
```diff
@@ -686,6 +686,18 @@
+        <div id="editorUndoBar" class="messageBar" hidden>
+          <div>
+            <div>
+              <span class="description" data-l10n-id="pdfjs-editor-undo-bar-text" data-l10n-args='{"type": "annotation"}'>[Annotation] removed</span>
```
I found out that localization wasn't working as expected but I hadn't realized that it was such a silly error... Thanks!

ryzokuken commented:
@Snuffleupagus thanks for the last slew of suggestions especially, it taught me a lot about the code and fixed the l10n bug. The code should ideally look much more manageable now. While I make more improvements and add tests, please let me know if you have any further comments.

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,63 @@
+  show(action, type) {
+    this.#hide();
+    this.#message.setAttribute("data-l10n-args", JSON.stringify({ type }));
+    this.#container.hidden = false;
```
```suggestion
    this.isOpen = true;
    this.#container.hidden = false;
```

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,63 @@
+
+  #container;
+
+  #controller = null;
```
```suggestion
  #controller = null;
  
  isOpen = false;
```

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,63 @@
+    this.#undoButton.focus();
+  }
+
+  #hide() {
+    this.#container.hidden = true;
```
```suggestion
  #hide() {
    if (!this.isOpen) {
      return;
    }
    this.isOpen = false;
    this.#container.hidden = true;
    
```

Snuffleupagus reviewed (commented):
Another thing that I think should be supported is closing the undo-bar with the <kbd>Esc</kbd> key, and the easiest solution would be to (after implementing the comments below) also add the following code *just after* [this existing code](https://github.com/mozilla/pdf.js/blob/689ffda9df29f24e03f6e5145a096584d7f166cc/web/app.js#L2963-L2970):
```javascript
if (this.editorUndoBar?.isOpen) {
  this.editorUndoBar.hide();
  handled = true;
}
```

ryzokuken commented:
I added the ESC key dismiss thingie but note that it's not part of the Figma spec which says:

> The message bar should persist on screen until:
> * User clicks “X” close
> * User creates a new annotation
> * User opens Print dialog
> * User opens Save dialog
> * User selects an option from the “more” menu
> * User closes tab
> * User closes session

Maybe we should get the ESC key condition added there.

Snuffleupagus commented:
> I added the ESC key dismiss thingie but note that it's not part of the Figma spec which says:
> 
> > The message bar should persist on screen until:
> > 
> > [...]

Note also the other cases mentioned previously in https://github.com/mozilla/pdf.js/pull/18900#pullrequestreview-2368482095, maybe those should be added to that list as well?

ryzokuken commented:
Absolutely, I'll mention this to the UX folks.

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,78 @@
+    this.#message = message;
+    this.#undoButton = undoButton;
+    this.#closeButton = closeButton;
+    this.#eventBus = eventBus;
```
There's *currently* no reason for introducing the `#eventBus` field, as far as I can tell, since you're only accessing it here in the constructor where it's directly available.

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,78 @@
+
+    // Caveat: we have to pick between registering these everytime the bar is
+    // shown and not having the ability to cleanup using AbortController.
+    this.#closeButton.addEventListener("click", this.#boundHide);
```
If you're moving this here, please just use the shorter local variable name.
```suggestion
    closeButton.addEventListener("click", this.#boundHide);
```

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,78 @@
+    // Caveat: we have to pick between registering these everytime the bar is
+    // shown and not having the ability to cleanup using AbortController.
+    this.#closeButton.addEventListener("click", this.#boundHide);
+    this.#eventBus.on("print", this.#boundHide);
+    this.#eventBus.on("download", this.#boundHide);
```
Again, let's use the local name whenever that's available.
Also, since this is internal usage please use the "private" `EventBus`-methods instead (note that all other code in the `web/` folder does this).
```suggestion
    eventBus._on("print", this.#boundHide);
    eventBus._on("download", this.#boundHide);
```

ryzokuken commented:
Good point, I cleaned up the class fields and simplified the constructor a bit.

nicolo-ribaudo commented:
The first commit also (accidentally) fixes a bug in the existing `.messageBar` usage, where the X was not vertically aligned in the button:

![image](https://github.com/user-attachments/assets/907d34a3-9118-4181-8bc3-726804112a8f)

Snuffleupagus commented on web/app.js:
```diff
@@ -451,6 +453,16 @@ const PDFViewerApplication = {
         : null;
     }
 
+    let editorUndoBar = null;
+    if (appConfig.editorUndoBar) {
+      editorUndoBar = new EditorUndoBar(
+        appConfig.editorUndoBar,
+        eventBus,
+        l10n
+      );
+    }
+    this.editorUndoBar = editorUndoBar;
```
This can now be simplified, see also the next comment.
```suggestion
    if (appConfig.editorUndoBar) {
      this.editorUndoBar = new EditorUndoBar(
        appConfig.editorUndoBar,
        eventBus
      );
    }
```

Snuffleupagus commented on web/app.js:
```diff
@@ -460,6 +472,7 @@ const PDFViewerApplication = {
       linkService: pdfLinkService,
       downloadManager,
       altTextManager,
+      editorUndoBar,
```
```suggestion
      editorUndoBar: this.editorUndoBar,
```

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,86 @@
+    freetext: "pdfjs-editor-undo-bar-message-freetext",
+    stamp: "pdfjs-editor-undo-bar-message-stamp",
+    ink: "pdfjs-editor-undo-bar-message-ink",
+    __multiple: "pdfjs-editor-undo-bar-message-multiple",
```
Won't a single underscore suffice here?

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,86 @@
+    eventBus._on("download", boundHide);
+  }
+
+  async show(undoAction, messageData) {
+    this.hide();
+    this.isOpen = true;
+    this.#controller = new AbortController();
+
+    this.#message.textContent =
+      typeof messageData === "string"
+        ? await this.#l10n.get(EditorUndoBar.#l10nMessages[messageData])
+        : await this.#l10n.get(EditorUndoBar.#l10nMessages.__multiple, {
+            count: messageData,
+          });
+    this.#container.hidden = false;
```
Why do we need to translate this "manually", since that should only be done as a last resort?
Also, the introduction of await here could lead to intermittent bugs with the bar reshowing after having been hidden elsewhere.

```suggestion
  show(undoAction, messageData) {
    this.hide();
    
    if (typeof messageData === "string") {
      this.#message.setAttribute("data-l10n-id", EditorUndoBar.#l10nMessages[messageData]);
    } else {
      this.#message.setAttribute("data-l10n-id", EditorUndoBar.#l10nMessages._multiple)
      this.#message.setAttribute("data-l10n-args", JSON.stringify({ count: messageData });
    }
    this.isOpen = true;
    his.#container.hidden = false;
    
    this.#controller = new AbortController();
```

nicolo-ribaudo replied:
Oh right, thank you!

Snuffleupagus commented on src/display/editor/annotation_editor_layer.js:
```diff
@@ -404,6 +404,9 @@ class AnnotationEditorLayer {
         return;
       }
+
+      this.#uiManager._eventBus.dispatch("annotationstart", { source: this });
```
*This comment applies elsewhere as well.*

 - Is there possibly a more "central" spot where this can be done, since having this code in a bunch of places seems potentially error prone (since it's easy to miss adding a case)?
 - The event name isn't very specific, as far as I'm concerned. (But please see below.)
 - Ideally the undo-bar hiding would, in my opinion, be placed within the `src/display/editor/`-code as much as possible. (With events only used for features that are clearly outside of that folder.)
   With that in mind, could we change this to the following instead?
   ```suggestion
         this.#uiManager._editorUndoBar?.hide()
   ```

nicolo-ribaudo replied:
> Is there possibly a more "central" spot where this can be done, since having this code in a bunch of places seems potentially error prone (since it's easy to miss adding a case)?

We originally tried looking for a central place but couldn't find a good one. The problem is some of the editors are created when they are "enabled" by clicking on the toolbar (example: the ink one), while others are created only when the user actually starts writing in them (example: the freetext one).

Snuffleupagus commented on src/display/editor/tools.js:
```diff
@@ -804,6 +807,7 @@ class AnnotationEditorUIManager {
       rotation: 0,
     };
     this.isShiftKeyDown = false;
+    this.#editorUndoBar = editorUndoBar;
```
To support the suggestion above.
```suggestion
    this._editorUndoBar = editorUndoBar;
```

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,95 @@
+    eventBus._on("beforeprint", boundHide);
+    eventBus._on("download", boundHide);
+    eventBus._on("annotationstart", boundHide);
+    eventBus._on("secondarytoolbaraction", boundHide);
```
While this suggestion would be slight abuse of notation, do we really need a new event just for this or could we use the already existing one instead?
```suggestion
    eventBus._on("reporttelemetry", evt => {
      if (evt?.details?.type === "buttons") {
        boundHide();
      }
    });
```

nicolo-ribaudo replied:
We originally considered this, but yeah it felt a bit too much abusing an event for something that's not meant to. Ideally an event called `"reporttelemetry"` should not have UI effects other than... reporting telemetry 😅.

Happy to change it if you feel strongly though.

calixteman replied:
Why not just add a click listener on the window (capture) and check if the target is under the toast bar ?
It's something we do in order to hide the secondary toolbar:
https://github.com/mozilla/pdf.js/blob/master/web/app.js#L2750-L2765

calixteman replied:
I meant the latter.
The main problem of having this list of event is that we've to "maintain" it, I mean if we add something in the future, we'll have to remember to update it.

Aditi-1400 committed:
Add tests for annotation delete popup

Snuffleupagus commented on test/integration/freetext_editor_spec.mjs:
```diff
@@ -3688,4 +3688,131 @@ describe("FreeText Editor", () => {
+      await closePages(pages);
+    });
+
+    it("must check that deleting a text editor can be undone using the undo button", async () => {
```
```suggestion
    it("must check that deleting a FreeText editor can be undone using the undo button", async () => {
```

Snuffleupagus commented on test/integration/freetext_editor_spec.mjs:
```diff
@@ -3688,4 +3688,131 @@ describe("FreeText Editor", () => {
     });
   });
+
+  describe("Undo annotation popup has the expected behaviour", () => {
```
*Similar comments apply elsewhere throughout the patch as well.*

```suggestion
  describe("Undo editor popup has the expected behaviour", () => {
```

Aditi-1400 replied:
I've changed this to `Undo deletion popup has the expected behaviour` across all files, since that seems more relevant, let me know if we can phrase it better.

Snuffleupagus commented on test/integration/freetext_editor_spec.mjs:
```diff
@@ -3688,4 +3688,131 @@ describe("FreeText Editor", () => {
+      );
+    });
+
+    it("Must check annotation popup displays the correct message", async () => {
```
This looks like an incomplete sentence.
```suggestion
    it("Must check that the undo editor popup displays the correct message", async () => {
```

Aditi-1400 replied:
Also changed this to `must check that the undo deletion describe the correct message` across all files.

Snuffleupagus commented on test/integration/highlight_editor_spec.mjs:
```diff
@@ -2144,4 +2145,276 @@ describe("Highlight Editor", () => {
     });
   });
+
+  describe("Annotation pop-up has the expected behaviour", () => {
```
Again, a strange test name. (And please be consistent, use *either* `popup` or `pop-up` consistently throughout the patch.)

Snuffleupagus commented on test/integration/highlight_editor_spec.mjs:
```diff
@@ -2144,4 +2145,276 @@ describe("Highlight Editor", () => {
+          await waitForSerialized(page, 0);
+          await page.waitForSelector("#editorUndoBar:not([hidden])");
+
+          await page.click('button[title="Undo"]');
```
This seems really brittle, and quite frankly unnecessary; please use the element id here.

Snuffleupagus commented on test/integration/highlight_editor_spec.mjs:
```diff
@@ -2144,4 +2145,276 @@ describe("Highlight Editor", () => {
+          await waitForSerialized(page, 0);
+          await page.waitForSelector("#editorUndoBar:not([hidden])");
+
+          await page.click('button[title="Undo"]');
```
This seems really brittle, and quite frankly unnecessary; please use the element id here.

Snuffleupagus commented on test/integration/highlight_editor_spec.mjs:
```diff
@@ -2144,4 +2145,276 @@ describe("Highlight Editor", () => {
+          await waitForSerialized(page, 0);
+          await page.waitForSelector("#editorUndoBar:not([hidden])");
+
+          await page.click('button[title="Tools"]');
```
This seems really brittle, and quite frankly unnecessary; please use the element id here.

Snuffleupagus commented on test/integration/ink_editor_spec.mjs:
```diff
@@ -453,4 +453,130 @@ describe("Ink Editor", () => {
+          await waitForSerialized(page, 0);
+
+          await page.waitForSelector("#editorUndoBar:not([hidden])");
+          await page.click('button[title="Undo"]');
```
This seems really brittle, and quite frankly unnecessary; please use the element id here.

Snuffleupagus commented on test/integration/stamp_editor_spec.mjs:
```diff
@@ -1281,4 +1281,98 @@ describe("Stamp Editor", () => {
+
+          await page.waitForSelector("#editorUndoBar:not([hidden])");
+
+          await page.click('button[title="Undo"]');
```
This seems really brittle, and quite frankly unnecessary; please use the element id here.

Snuffleupagus commented on test/integration/freetext_editor_spec.mjs:
```diff
@@ -3688,4 +3688,131 @@ describe("FreeText Editor", () => {
+              document
+                .querySelector("#editorUndoBarMessage")
+                .textContent.trim() !== "",
+            { timeout: 1000 }
```
*Here, and in lots of other places throughout the patch:*

Do *not* use arbitrary timeouts like this, since that leads to intermittent test failures.

Snuffleupagus reviewed (changes requested):
There's a few additional cases where the undo-bar should also be hidden (which should be possible to implement within the `src/display/editor/`-code):
 - When toggling editing off (i.e. setting `AnnotationEditorType.NONE`).
 - When destroying the editor.

Snuffleupagus commented on test/integration/highlight_editor_spec.mjs:
```diff
@@ -2144,4 +2145,276 @@ describe("Highlight Editor", () => {
+      );
+    });
+
+    it("must check that the popup disappears when an option from the 'more' menu is selected", async () => {
```
```suggestion
    it("must check that the popup disappears when an option from the secondaryToolbar is used", async () => {
```

Snuffleupagus commented on test/integration/freetext_editor_spec.mjs:
```diff
@@ -3688,4 +3688,130 @@ describe("FreeText Editor", () => {
+            const messageElement = document.querySelector(
+              "#editorUndoBarMessage"
+            );
+            return messageElement && messageElement.textContent.trim() !== "";
```
Here, and elsewhere throughout the patch:
```suggestion
            return messageElement?.textContent.trim() !== "";
```

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,101 @@
+    eventBus._on("beforeprint", boundHide);
+    eventBus._on("download", boundHide);
+    eventBus._on("secondarytoolbaraction", boundHide);
+    eventBus._on("annotationeditormodechanged", ({ mode }) => {
+      if (mode === AnnotationEditorType.NONE) {
+        this.hide();
+      }
+    })
```
As mentioned in https://github.com/mozilla/pdf.js/pull/18900#pullrequestreview-2379266564, can't this be added directly in the relevant `src/display/editor/`-code so that we don't need an event here!?
The following method seems relevant: https://github.com/mozilla/pdf.js/blob/d37e4b08e4006b382b14b1596c5f70df7e93fbd9/src/display/editor/tools.js#L1616-L1671

Snuffleupagus replied:
Similarly, you also want to hide the undo-bar in the following method: https://github.com/mozilla/pdf.js/blob/d37e4b08e4006b382b14b1596c5f70df7e93fbd9/src/display/editor/tools.js#L864-L892

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,101 @@
+ * limitations under the License.
+ */
+
+import { AnnotationEditorType } from "../src/shared/util.js";
```
Sorry, but you cannot import like this in the `web/`-folder since that could end up pulling in a very significant chunk of code in the *built* `viewer.mjs` file.

For an example of how this *should* be done, see [one example here](https://github.com/mozilla/pdf.js/blob/d37e4b08e4006b382b14b1596c5f70df7e93fbd9/web/pdf_viewer.js#L31-L39), however note that given the comment above this import shouldn't even be necessary here.

ryzokuken commented:
Thanks for all the insightful suggestions @Snuffleupagus. Just cleaned up the commit history somewhat and things should be looking better now, please take a look.

Snuffleupagus commented on web/secondary_toolbar.js:
```diff
@@ -219,6 +219,7 @@ class SecondaryToolbar {
     // All items within the secondary toolbar.
     for (const { element, eventName, close, eventDetails } of buttons) {
       element.addEventListener("click", evt => {
+        eventBus.dispatch("secondarytoolbaraction", { source: this });
```
If we're going to introduce a new event here (which I'm not crazy about personally), then at the very least it should be placed *after* the "reporttelemetry"-dispatch below.

ryzokuken replied:
I'll do this and ask others later today how they prefer we do this.

Snuffleupagus reviewed (changes requested):
Since "Break messageBar class out" is a separate commit, and that does make sense, you still haven't fixed the following comments in the *correct* patch:
 - https://github.com/mozilla/pdf.js/pull/18900#discussion_r1799814834
 - https://github.com/mozilla/pdf.js/pull/18900#discussion_r1799817954

ryzokuken committed:
Pop open a message when user deletes an annotation

When a user deletes any number of annotations, they are notified of the action
by a popup message with an undo button. Besides that, this change reuses the
existing messageBar CSS class from the new alt-text dialog as much as possible.

flodolo commented on l10n/en-US/viewer.ftl:
```diff
@@ -493,3 +493,20 @@ pdfjs-editor-alt-text-settings-editor-title = Alt text editor
+pdfjs-editor-undo-bar-message-stamp = Image removed
+# Variables:
+#   $count (Number) - the number of removed annotations.
+pdfjs-editor-undo-bar-message-multiple = { $count } annotations removed
```
```suggestion
pdfjs-editor-undo-bar-message-multiple = 
    { $count ->
        [one] { $count } annotation removed
       *[other] { $count } annotations removed
    }
```

This needs a plural form.

ryzokuken replied:
@flodolo the message is actually never rendered unless the number is two or more but I imagine it'll make it easier to port it to other languages with more plural forms?

flodolo replied:
Other languages have more [complex plural rules](https://www.unicode.org/cldr/charts/45/supplemental/language_plural_rules.html) than English (up to 6 forms). We want to instruct tools that this string needs plural support, even if the `one` category is never used for English.

ryzokuken replied:
fair enough, done!

Snuffleupagus commented on web/app.js:
```diff
@@ -2764,6 +2771,20 @@ function onClick(evt) {
+    return;
+  }
+  if (this.appConfig.secondaryToolbar?.toolbar.contains(evt.target)) {
+    this.editorUndoBar?.hide();
```
Given the early return above, you can't reach this line unless the undo-bar is available.
```suggestion
    this.editorUndoBar.hide();
```

Snuffleupagus commented:
https://github.com/mozilla/pdf.js/pull/18900#pullrequestreview-2368482095 mentioned undoing via the keyboard, in which case I believe that we also should hide the undo-bar, however looking at the code that may not have been implemented yet (but please note that I've not actually tested the patch).

Would it work if we always invoke `this._editorUndoBar?.hide();` in the following method?
https://github.com/mozilla/pdf.js/blob/d37e4b08e4006b382b14b1596c5f70df7e93fbd9/src/display/editor/tools.js#L2010-L2020

ryzokuken commented:
After discussion, I've made the undo bar be dismissed on _any_ keypress that isn't shift and/or tab.

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -39,10 +39,7 @@ class EditorUndoBar {
-    const boundHide = this.hide.bind(this);
-    closeButton.addEventListener("click", boundHide);
-    eventBus._on("beforeprint", boundHide);
-    eventBus._on("download", boundHide);
```
Removing these listeners means that a number of ways in which printing and downloading can be triggered will no longer hide the undo-bar.

nicolo-ribaudo replied:
Ctrl+P and Ctrl+S are handled by the key press handler. Are there other ways that involve neither the keyboard nor the toolbar?

nicolo-ribaudo replied:
Assuming that https://github.com/mozilla/pdf.js/pull/18900#discussion_r1806759172 gets done @ryzokuken

nicolo-ribaudo replied:
Nevermind, clicking on the "print" icon in the toolbar does indeed not close the bar.

ryzokuken replied:
ah shucks atm only clicking on the secondary toolbar dismisses the undo bar. What should I do? Re-add the event listeners or also make the bar go away when someone clicks the main toolbar?

nicolo-ribaudo replied:
After playing around with it, it seems like when using the built-in viewer in firefox (but not when using https://mozilla.github.io/pdf.js/web/viewer.html) you can also save and print using the browser's menu.

I think it's thus better to keep these two event listeners.

ryzokuken replied:
re-added, thanks both

Snuffleupagus commented on web/app.js:
```diff
@@ -3125,6 +3121,10 @@ function onKeyDown(evt) {
     evt.preventDefault();
   }
+
+  if (this.editorUndoBar?.isOpen && evt.keyCode !== 9 && evt.keyCode !== 16) {
```
Please add comments explaining what these keyCodes are, note the rest of this function.

Snuffleupagus commented on web/app.js:
```diff
@@ -3125,6 +3121,10 @@ function onKeyDown(evt) {
   }
+
+  if (this.editorUndoBar?.isOpen && evt.keyCode !== 9 && evt.keyCode !== 16) {
+    this.editorUndoBar?.hide();
```
Again, optional chaining isn't necessary here given the check above.

Snuffleupagus commented on web/app.js:
```diff
@@ -3125,6 +3121,10 @@ function onKeyDown(evt) {
     evt.preventDefault();
   }
+
+  if (this.editorUndoBar?.isOpen && evt.keyCode !== 9 && evt.keyCode !== 16) {
+    this.editorUndoBar?.hide();
+  }
```
Placing this code at the end might miss some cases, please note the early return here: https://github.com/mozilla/pdf.js/blob/bde36f28bee3a01495375ae88d6ac2e3bf217c74/web/app.js#L2898-L2904

nicolo-ribaudo commented on web/viewer.html:
```diff
@@ -686,6 +686,20 @@
 <!--#endif-->
       </div>  <!-- dialogContainer -->
 
+      <div id="editorUndoBar" class="messageBar" hidden>
```
This should have `role="status"` according to the Figma doc.

Snuffleupagus commented:
/botio integrationtest

Snuffleupagus commented:
/botio-linux preview

Snuffleupagus reviewed (changes requested):
Unfortunately there seems to be various issues related to styling, the following are ones I spotted in very quick testing:
 - The undo-bar layout doesn't support RTL locales properly, since it seems that various margin/padding entries aren't using logical values to make them direction aware.
   
   ![rtl](https://github.com/user-attachments/assets/299be9bd-845c-4717-bcbd-769f1c2fb15b)

 - The "Undo" and "Close" buttons have quite different focus outlines, both in size *and* styling, which looks inconsistent/strange.
   (Also, the "Close" button seems to have the same styling for focus and hover thus making those states difficult to tell apart.)
   
   ![undo](https://github.com/user-attachments/assets/c2134d59-0cb2-4c02-bee4-d0ae6a476117)

   ![close](https://github.com/user-attachments/assets/5638cc27-3a58-43de-a325-a5260977d8a5)

 - The "message"-span has selectable text, with a text-cursor when hovering it, which I don't believe is intended.

 - The undo-bar appears *under* the sidebar when the viewer is narrow enough and/or when the sidebar is resized.

   ![sidebar](https://github.com/user-attachments/assets/b9fed568-a8fe-403c-9148-4ed91b1bb45c)

calixteman commented on src/display/editor/tools.js:
```diff
@@ -889,6 +893,7 @@ class AnnotationEditorUIManager {
       clearTimeout(this.#translationTimeoutId);
       this.#translationTimeoutId = null;
     }
+    this._editorUndoBar?.hide();
```
Can't we make a better clean-up ?
The class `EditorUndoBar` should have a method `destroy` to make a full clean-up.

Snuffleupagus replied:
Sorry, but what do you mean with "full clean-up" here?

Please keep in mind that the GENERIC viewer supports opening more than one PDF document, so we cannot completely destroy everything related to the undo-bar without things breaking.

calixteman replied:
I was meaning removing the event listeners created in the ctor which can be re-added if the bar is shown again.

Snuffleupagus replied:
It seems that https://github.com/mozilla/pdf.js/pull/18900#discussion_r1822997001 wasn't taken into account when implementing this!?

Snuffleupagus replied:
Basically, you want something like the following diff which also has the advantage of registering more stuff lazily.
```diff
diff --git a/web/editor_undo_bar.js b/web/editor_undo_bar.js
index 7bb96f65f..36b967640 100644
--- a/web/editor_undo_bar.js
+++ b/web/editor_undo_bar.js
@@ -14,11 +14,13 @@
  */
 
 class EditorUndoBar {
+  #closeButton;
+
   #container;
 
-  #controller = null;
+  #eventBus;
 
-  #instController = null;
+  #initController = null;
 
   isOpen = false;
 
@@ -26,6 +28,8 @@ class EditorUndoBar {
 
   #undoButton;
 
+  #showController = null;
+
   static #l10nMessages = Object.freeze({
     highlight: "pdfjs-editor-undo-bar-message-highlight",
     freetext: "pdfjs-editor-undo-bar-message-freetext",
@@ -38,21 +42,27 @@ class EditorUndoBar {
     this.#container = container;
     this.#message = message;
     this.#undoButton = undoButton;
-
-    this.#instController = new AbortController();
-    const boundHide = this.hide.bind(this);
-    const opts = { signal: this.#instController.signal };
-    closeButton.addEventListener("click", boundHide, opts);
-    eventBus._on("beforeprint", boundHide, opts);
-    eventBus._on("download", boundHide, opts);
+    this.#closeButton = closeButton;
+    this.#eventBus = eventBus;
   }
 
   destroy() {
-    this.#instController.abort();
+    this.#initController?.abort();
+    this.#initController = null;
+
     this.hide();
   }
 
   show(undoAction, messageData) {
+    if (!this.#initController) {
+      this.#initController = new AbortController();
+      const opts = { signal: this.#initController.signal };
+
+      const boundHide = this.hide.bind(this);
+      this.#closeButton.addEventListener("click", boundHide, opts);
+      this.#eventBus._on("beforeprint", boundHide, opts);
+      this.#eventBus._on("download", boundHide, opts);
+    }
     this.hide();
 
     if (typeof messageData === "string") {
@@ -73,7 +83,7 @@ class EditorUndoBar {
     this.isOpen = true;
     this.#container.hidden = false;
 
-    this.#controller = new AbortController();
+    this.#showController = new AbortController();
 
     this.#undoButton.addEventListener(
       "click",
@@ -81,7 +91,7 @@ class EditorUndoBar {
         undoAction();
         this.hide();
       },
-      { signal: this.#controller.signal }
+      { signal: this.#showController.signal }
     );
     this.#undoButton.focus();
   }
@@ -92,8 +102,9 @@ class EditorUndoBar {
     }
     this.isOpen = false;
     this.#container.hidden = true;
-    this.#controller?.abort();
-    this.#controller = null;
+
+    this.#showController?.abort();
+    this.#showController = null;
   }
 }
```

ryzokuken replied:
@Snuffleupagus done, PTAL

calixteman commented on src/display/editor/freetext.js:
```diff
@@ -550,6 +550,8 @@ class FreeTextEditor extends AnnotationEditor {
       return this.div;
     }
 
+    this._uiManager._editorUndoBar?.hide();
```
Here and in the other types of editors, isn't it possible to get rid of this ?

ryzokuken replied:
All editors except "ink" call this in the `render` method. If you're comfortable with me removing that special case, I could move this to `AnnotationEditor#render`.

ryzokuken replied:
@calixteman done with the caveat mentioned above (in case of the ink editor, the bar is dismissed when it's selected from the toolbar and not when it's actually created). Check it out and let me know what you think.

ryzokuken replied:
Unmarking this comment as unresolved so you could consider this before the final go-ahead.

nicolo-ribaudo replied:
To be clear, now if you:
- draw an ink annotation
- delete it
- draw another one

the popup bar does not close.

ryzokuken replied:
This is fixed by moving the hide action to the parent class while keeping the special case for ink.js, thanks for the pointer @calixteman

calixteman commented:
In HCM, the Undo button has no outline.

ryzokuken commented:
@calixteman fixed those issues and squashed the commits, should be looking better now.

@Snuffleupagus could you please take a look as well?

Thanks both!

nicolo-ribaudo commented:
/botio-linux preview

calixteman commented:
I just tested and the height of the toast isn't correct.

calixteman commented:
> I just tested and the height of the toast isn't correct.

The wrong height is a regression from https://bugzilla.mozilla.org/show_bug.cgi?id=1920160.

ryzokuken commented:
@Snuffleupagus could you please take a look again? Thanks

Snuffleupagus commented:
> @Snuffleupagus could you please take a look again? Thanks

Sorry, but to save my time I'm now going to defer any further reviewing until this is *fully* done (and "fixup" patches folded into their parent comments) with *all* previous review comments addressed and *everything* working correctly.

For example, it seems that https://github.com/mozilla/pdf.js/pull/18900#pullrequestreview-2399239975 is still open given the issue mentioned in https://github.com/mozilla/pdf.js/pull/18900#discussion_r1827487927?

ryzokuken commented:
Thanks for the reminder @Snuffleupagus I discussed the open issue with @calixteman, addressed it and squashed all the commits together\.

calixteman commented:
/botio-linux preview

Snuffleupagus commented on web/dialog.css:
```diff
@@ -293,7 +293,8 @@
 
       > div {
-        &::before, > div {
+        &::before,
+        > div {
```
Sorry, but why was this changed since it looks unrelated to the patch?

nicolo-ribaudo replied:
Previously `.messageBar` had more vertical padding, that @ryzokuken had to reduce because of the `Undo` button. This `margin-block` preserves the old spacing in the `.messageBar` usage for the new alt text dialog.

However, the one at line 307 is not needed since it's just duplicating this one.

ryzokuken replied:
@Snuffleupagus I overlooked that, now the first commit has the reformatting and doesn't fail the linter.

Snuffleupagus commented on web/message_bar.css:
```diff
@@ -124,3 +124,97 @@
+#editorUndoBar {
+  --text-primary-color: #15141a;
+
+  --message-bar-icon: url(images/secondaryToolbarButton-documentProperties.svg);
```
Given where the current CSS file is being bundled, in order to prevent breakage of the `gulp components` build this image needs to be added to the following list.
https://github.com/mozilla/pdf.js/blob/9bf9bbda0b828f528679a981eb416b805880460e/gulpfile.mjs#L1147-L1155

ryzokuken replied:
Great catch, I had no clue. Just fixed.

Snuffleupagus commented:
/botio integrationtest

ryzokuken commented:
Changed the position of the popup from the bottom to the top of the document (right below the toolbar) in order to match the latest feedback from the design systems team.

calixteman commented:
/botio-linux preview

calixteman commented on web/message_bar.css:
```diff
@@ -0,0 +1,220 @@
+  padding-block: 8px;
+  padding-inline: 16px 8px;
+
+  font-family: sans-serif;
```
nit: `font: menu`

calixteman commented on web/message_bar.css:
```diff
@@ -0,0 +1,220 @@
+  justify-content: center;
+  align-items: center;
+  gap: 8px;
+
```
nit: disable the text selection in the toast.

calixteman commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,112 @@
+    _multiple: "pdfjs-editor-undo-bar-message-multiple",
+  });
+
+  constructor({ container, message, undoButton, closeButton }, eventBus) {
```
nit: disable the context menu in the toast (see `noContextMenu` in `pdfjs-lib`.

calixteman commented:
On Windows, when the toast appears, NVDA is reading "Toolbar  Undo button" which means that the description and the close button are missed: they probably needs to have some aria attributes.
I didn't check on mac with VoiceOver but you should give it a try to make sure that everything is correctly read.

ryzokuken commented:
@calixteman fixed the ARIA issue and tested on VO, will test on NVDA tomorrow. We had to add a setTimeout for the focus in order to get VO to switch to that instead of just repeating the document title.

ryzokuken commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,124 @@
+    // Without the setTimeout, VoiceOver will read out the document title
+    // instead of the popup label.
+    setTimeout(() => {
+      this.#container.focus();
```
Great catch, thanks!

Snuffleupagus replied:
This comment was made obsolete by instead implementing https://github.com/mozilla/pdf.js/pull/18900#discussion_r1858219587.

ryzokuken commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,124 @@
+      },
+      { signal: this.#showController.signal }
+    );
+    this.#container.addEventListener(
+      "contextmenu",
+      noContextMenu,
+      this.#showController.signal
+    );
```
I put it here to use the abortcontroller with it but if that's not as big a deal I'm happy to move it above.

ryzokuken commented:
@Snuffleupagus thanks for your comments again, I just addressed them, hopefully it should be good to go now.

@calixteman I just tested the screen reader on Windows and it works as expected!

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,122 @@
+
+    // Without the setTimeout, VoiceOver will read out the document title
+    // instead of the popup label.
+    setTimeout(() => {
+      if (this.isOpen) {
+        this.#container.focus();
+      }
+    }, 100);
```
Also, it occurred to me that this we should instead just remove the timeout when hiding the undo-bar to prevent any issues:
```diff
diff --git a/web/editor_undo_bar.js b/web/editor_undo_bar.js
index 50f6ec2c5..7a260590e 100644
--- a/web/editor_undo_bar.js
+++ b/web/editor_undo_bar.js
@@ -22,6 +22,8 @@ class EditorUndoBar {
 
   #eventBus = null;
 
+  #focusTimeout = null;
+
   #initController = null;
 
   isOpen = false;
@@ -100,10 +102,9 @@ class EditorUndoBar {
 
     // Without the setTimeout, VoiceOver will read out the document title
     // instead of the popup label.
-    setTimeout(() => {
-      if (this.isOpen) {
-        this.#container.focus();
-      }
+    this.#focusTimeout = setTimeout(() => {
+      this.#container.focus();
+      this.#focusTimeout = null;
     }, 100);
   }
 
@@ -116,6 +117,11 @@ class EditorUndoBar {
 
     this.#showController?.abort();
     this.#showController = null;
+
+    if (this.#focusTimeout) {
+      clearTimeout(this.#focusTimeout);
+      this.#focusTimeout = null;
+    }
   }
 }
```

ryzokuken commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,122 @@
+      const opts = { signal: this.#initController.signal };
+      const boundHide = this.hide.bind(this);
+
+      this.#closeButton.addEventListener("click", boundHide, opts);
+      this.#eventBus._on("beforeprint", boundHide, opts);
+      this.#eventBus._on("download", boundHide, opts);
+      this.#container.addEventListener("contextmenu", noContextMenu);
```
Thanks!

ryzokuken commented:
@Snuffleupagus how does it look now?

calixteman commented on test/integration/freetext_editor_spec.mjs:
```diff
@@ -3688,4 +3688,130 @@ describe("FreeText Editor", () => {
+          const rect = await getRect(page, ".annotationEditorLayer");
+          const data = "Hello PDF.js World !!";
+          await page.mouse.click(rect.x + 100, rect.y + 100);
+          await page.waitForSelector(getEditorSelector(0), {
```
nit: just set a variable `const editorSelector = getEditorSelector(0);` and use it elsewhere in the test (the same comment applies for the other tests).

Snuffleupagus commented:
It's not possible to use the keyboard to "click" on the Undo-button, since tabbing to it and then pressing either <kbd>Enter</kbd> or <kbd>Space</kbd> doesn't work.

calixteman commented:
@ryzokuken could you fix the bug mentioned in https://github.com/mozilla/pdf.js/pull/18900#issuecomment-2504592921 ? and add a test (just for one kind of editor) to check that "clicking" on undo with the keyboard is working as expected. Thank you.

ryzokuken commented:
@calixteman okay should be fixed now, also addressed your suggestion about moving out `editorSelector`.

Snuffleupagus commented:
> okay should be fixed now,

Unfortunately that's not the case, since there's various problems with the latest update:
 - `Spacebar` can also be used to "click" with the keyboard.
 - Ignoring those keys in the `onKeyDown` handler means that if focus is *outside* of the undo-bar then it won't be closed as intended in all situations.

```diff
diff --git a/web/app.js b/web/app.js
index 3b4e0d8db..945c9e76a 100644
--- a/web/app.js
+++ b/web/app.js
@@ -2780,14 +2780,9 @@ function onKeyUp(evt) {
 function onKeyDown(evt) {
   this._isCtrlKeyDown = evt.key === "Control";
 
-  if (
-    this.editorUndoBar?.isOpen &&
-    evt.keyCode !== 9 &&
-    evt.keyCode !== 13 &&
-    evt.keyCode !== 16
-  ) {
-    // Hide undo bar on keypress except for Shift, Tab , Shift+Tab or Enter
-    this.editorUndoBar.hide();
+  if (this.editorUndoBar?.isOpen && evt.keyCode !== 9 && evt.keyCode !== 16) {
+    // Hide undo bar on keypress except for Shift, Tab, or Shift+Tab.
+    this.editorUndoBar.hide(/* isFromKeyboard = */ true);
   }
 
   if (this.overlayManager.active) {
diff --git a/web/editor_undo_bar.js b/web/editor_undo_bar.js
index aa064e384..ef33ce2ea 100644
--- a/web/editor_undo_bar.js
+++ b/web/editor_undo_bar.js
@@ -13,6 +13,7 @@
  * limitations under the License.
  */
 
+import { getActiveOrFocusedElement } from "./ui_utils.js";
 import { noContextMenu } from "pdfjs-lib";
 
 class EditorUndoBar {
@@ -108,10 +109,18 @@ class EditorUndoBar {
     }, 100);
   }
 
-  hide() {
+  hide(isFromKeyboard = false) {
     if (!this.isOpen) {
       return;
     }
+    if (isFromKeyboard) {
+      const curElement = getActiveOrFocusedElement();
+      if (this.#undoButton.contains(curElement)) {
+        // Ensure that "clicking" on the `undoButton` with the keyboard will
+        // actually undo the editor.
+        return;
+      }
+    }
     this.isOpen = false;
     this.#container.hidden = true;
```

ryzokuken commented:
Fixed the bug, added tests to check for that specific behavior and rebased again. How does it look?

Snuffleupagus commented on src/display/editor/annotation_editor_layer.js:
```diff
@@ -386,6 +386,7 @@ class AnnotationEditorLayer {
         // Do nothing on right click.
         return;
       }
+
```
This looks like an unrelated change, please revert.

Snuffleupagus commented on web/app.js:
```diff
@@ -2759,6 +2780,18 @@ function onKeyUp(evt) {
+    this.editorUndoBar?.isOpen &&
+    evt.keyCode !== 9 &&
+    evt.keyCode !== 16 &&
+    (getActiveOrFocusedElement() !== this.editorUndoBar.undoButton ||
```
Similar to other code in this file, can't we do the following here (and avoid having to introduce a getter just for this one spot)?
```suggestion
    (getActiveOrFocusedElement() !== this.appConfig.editorUndoBar.undoButton ||
```

Snuffleupagus commented on web/app.js:
```diff
@@ -2759,6 +2780,18 @@ function onKeyUp(evt) {
+    (getActiveOrFocusedElement() !== this.editorUndoBar.undoButton ||
+      !(evt.keyCode === 13 || evt.keyCode === 32))
+  ) {
+    // Hide undo bar on keypress except for Shift, Tab , Shift+Tab
+    // Also avoid hiding if the undo button is triggered
```
To aid readability, please remember to end sentences in comments with a period.
```suggestion
    // Hide undo bar on keypress except for Shift, Tab, Shift+Tab.
    // Also avoid hiding if the undo button is triggered.
```

Snuffleupagus commented on web/editor_undo_bar.js:
```diff
@@ -0,0 +1,132 @@
+      this.#focusTimeout = null;
+    }
+  }
+
+  get undoButton() {
+    return this.#undoButton;
+  }
```
This getter is unnecessary with previous comments fixed.

Snuffleupagus commented on test/integration/highlight_editor_spec.mjs:
```diff
@@ -2165,4 +2171,461 @@ describe("Highlight Editor", () => {
+          await page.waitForSelector("#editorUndoBar:not([hidden])");
+
+          await page.focus("#editorUndoBarUndoButton"); // we have to simulate focus like this to avoid the wait
+          await page.keyboard.press("Enter");
+          await waitForSerialized(page, 1);
+          await page.waitForSelector(editorSelector);
+          await page.waitForSelector(
+            `.page[data-page-number = "1"] svg.highlight[fill = "#FFFF00"]`
+          );
+        })
+      );
+    });
+
+    it("must work properly when selecting undo by keyboard using space key", async () => {
+      await Promise.all(
+        pages.map(async ([browserName, page]) => {
+          await switchToHighlight(page);
+
+          const rect = await getSpanRectFromText(page, 1, "Abstract");
+          const x = rect.x + rect.width / 2;
+          const y = rect.y + rect.height / 2;
+          await page.mouse.click(x, y, { count: 2, delay: 100 });
+          await page.waitForSelector(editorSelector);
+          await waitForSerialized(page, 1);
+
+          await page.waitForSelector(`${editorSelector} button.delete`);
+          await page.click(`${editorSelector} button.delete`);
+          await waitForSerialized(page, 0);
+          await page.waitForSelector("#editorUndoBar:not([hidden])");
+
+          await page.focus("#editorUndoBarUndoButton"); // we have to simulate focus like this to avoid the wait
+          await page.keyboard.press(" ");
+          await waitForSerialized(page, 1);
+          await page.waitForSelector(editorSelector);
+          await page.waitForSelector(
+            `.page[data-page-number = "1"] svg.highlight[fill = "#FFFF00"]`
+          );
+        })
+      );
+    });
```
Do we really need *two separate* tests for this, or can they be combined into just one test that checks both Enter and Space one after another?

Snuffleupagus commented on web/app.js:
```diff
@@ -2759,6 +2780,18 @@ function onKeyUp(evt) {
+    this.editorUndoBar?.isOpen &&
+    evt.keyCode !== 9 &&
+    evt.keyCode !== 16 &&
+    (getActiveOrFocusedElement() !== this.editorUndoBar.undoButton ||
+      !(evt.keyCode === 13 || evt.keyCode === 32))
```
Also, these two lines should probably be swapped to do the "cheaper" `keyCode` checks *first* to avoid having to unconditionally call `getActiveOrFocusedElement` here.

ryzokuken commented:
@Snuffleupagus thanks for the comments! Fixed now.

Snuffleupagus commented:
/botio integrationtest

Snuffleupagus commented on web/app.js:
```diff
@@ -2759,6 +2780,18 @@ function onKeyUp(evt) {
+    this.editorUndoBar?.isOpen &&
+    evt.keyCode !== 9 &&
+    evt.keyCode !== 16 &&
+    (!(evt.keyCode === 13 || evt.keyCode === 32) ||
+      getActiveOrFocusedElement() !== this.appConfig.editorUndoBar.undoButton)
```
Nit: This could be "simplified" a little bit by moving the negation outside.
```suggestion
    !((evt.keyCode === 13 || evt.keyCode === 32) &&
      getActiveOrFocusedElement() === this.appConfig.editorUndoBar.undoButton)
```

ryzokuken replied:
okay, done!

Snuffleupagus reviewed (approved):
r=me, with one more suggestion.

calixteman reviewed (approved):
LGTM. Thank you.
